### PR TITLE
Expand PaddedGraphGenerator to generate pairs (or more) of graphs

### PR DESCRIPTION
**Pull request by huonw**
This PR generalises `PaddedGraphGenerator` to support taking a sequence of rows of one or more of graphs (either ilocs or `StellarGraph` instances). It yields the padded features, mask and adjacency matrix for each element, as tensors in succession.

For instance, if the input looks like `[(12, 23)]` the output feature tensors will contain six elements:

1. padded features for graph 12
2. padding mask for graph 12
3. padded adjacency matrix for graph 12
4. padded features for graph 23
5. padding mask for graph 23
6. padded adjacency matrix for graph 23.

This allows implementing unsupervised graph classification algorithms, like [UGraphEmb](http://arxiv.org/abs/1904.01098) (in theory). The tensors being in a group for each batch means the input for a model is just the concatenation of the input tensors from multiple calls to `in_out_tensors`, e.g.

```python
generator = PaddedGraphGenerator(...)
model = ...

inp1, out1 = model.in_out_tensors()
inp2, out2 = model.in_out_tensors()

combined = merge_outputs_somehow(out1, out2)
model = tensorflow.keras.Model(inp1 + inp2, combined)
model.compile(...)

train_gen = generator.flow([(0, 1), (1, 2)], targets=[1.23, 4.56])
model.fit(train_gen)
```

This PR also adds a basic demo notebook inspired by UGraphEmb, running a graph embedding task on the PROTEINS dataset. The notebook is added to `demos/embeddings`, but might be better suited to another directory?

See: #1523

### CODE REVIEW

**Comment by timpitman** [comment]
> The notebook is added to `demos/embeddings`, but might be better suited to another directory?

we have a 'graph-classification' directory - that might be more appropriate?

File: demos/embeddings/gcn-unsupervised-graph-embeddings.ipynb [hunk 1]
```diff
@@ -0,0 +1,737 @@
+{
```
Comment (kjun9) [inline demos/embeddings/gcn-unsupervised-graph-embeddings.ipynb:1]: <blockquote>This demo demonstrated training a graph classification model without supervision. This model could be used to compute embedding vectors or representations for graphs.</blockquote><p>Is it technically more like "unsupervsed representation learning on graphs" + "supervised graph classification", rather than referring to the whole demo as being unsupervised? It looks like a supervised classification workflow towards the bottom if I understood correctly</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/stellargraph/stellargraph/pull/1626/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

**Comment by timpitman** [comment]
Are we implementing enough of UGraphEmb to list it in our main algorithms?

File: demos/embeddings/gcn-unsupervised-graph-embeddings.ipynb (hunk 1 above)
Comment (huonw) [reply demos/embeddings/gcn-unsupervised-graph-embeddings.ipynb:1]: This follows our demos for other unsupervised node classification tasks, where representations are computed without supervision and then there's example downstream tasks which may or may not be supervised. For example, unsupervised graphsage: https://stellargraph.readthedocs.io/en/latest/demos/embeddings/graphsage-unsupervised-sampler-embeddings.html#Downstream-task

I'll tweak the heading to hopefully make this clearer, replacing "Evaluate the model" with "Downstream tasks".

File: stellargraph/mapper/padded_graph_generator.py [hunk 2]
```diff
@@ -139,13 +139,30 @@ def flow(
...
 
-        if isinstance(graphs[0], StellarGraph):
-            self._check_graphs(graphs)
+        graphs_array = np.asarray(graphs)
+
+        if len(graphs_array.shape) == 1:
+            graphs_array = graphs_array[:, None]
+        elif len(graphs_array.shape) != 2:
+            raise ValueError(
+                f"graphs: expected a shape of length 1 or 2, found shape {graphs_aray.shape}"
```
[the remark this answers is not in the document]
Comment (huonw) [reply stellargraph/mapper/padded_graph_generator.py]: Good catch. I added a test.

File: stellargraph/mapper/padded_graph_generator.py [hunk 3]
```diff
@@ -139,13 +139,30 @@ def flow(
...
+        if isinstance(flat_graphs[0], StellarGraph):
+            self._check_graphs(flat_graphs)
+            graphs = flat_graphs
+            selected_ilocs = np.arange(len(graphs)).reshape(graphs_array.shape)
         else:
-            graphs = [self.graphs[i] for i in graphs]
+            selected_ilocs = graphs_array
+            graphs = self.graphs
+
+        # import pdb; pdb.set_trace()
```
[the remark this answers is not in the document]
Comment (huonw) [reply stellargraph/mapper/padded_graph_generator.py]: Oops!

**Comment by huonw** [comment]
> we have a graph-classification directory - that might be more appropriate?

I'm not sure. I considered that initially, but I felt that it didn't fit into the existing categories of demos and the table in `embeddings/` was a better fit. Maybe that one is better though.

> Are we implementing enough of UGraphEmb to list it in our main algorithms?

I don't think so. This implements the training procedure, but the actual model is a bit more complicated and something I don't think we have yet. I think it will be some sort of attention pooling for the graphs (in addition to our current sort pooling and mean/sum pooling) and maybe a tweak to how embeddings are concatenated:

image
image

I filed #1627 that can cover more of this.

(Thanks for the fast reviews!)
